packages/next/src/client/react-client-callbacks/app-router.ts

```diff
@@ -0,0 +1,85 @@
+// This file is only used in app router due to the specific error state handling.
+
+import type { HydrationOptions } from 'react-dom/client'
+import { getReactStitchedError } from '../components/react-dev-overlay/internal/helpers/stitched-error'
+import { handleClientError } from '../components/react-dev-overlay/internal/helpers/use-error-handler'
+import { isNextRouterError } from '../components/is-next-router-error'
+import { isBailoutToCSRError } from '../../shared/lib/lazy-dynamic/bailout-to-csr'
+import { reportGlobalError } from './report-global-error'
+import isError from '../../lib/is-error'
+import { originConsoleError } from '../components/globals/intercept-console-error'
+
+export const onCaughtError: HydrationOptions['onCaughtError'] = (
+  err,
+  errorInfo
+) => {
+  // Skip certain custom errors which are not expected to be reported on client
+  if (isBailoutToCSRError(err) || isNextRouterError(err)) return
+
+  const stitchedError = getReactStitchedError(err)
+
+  if (process.env.NODE_ENV === 'development') {
+    const errorBoundaryComponent = errorInfo?.errorBoundary?.constructor
+    const errorBoundaryName =
+      // read react component displayName
+      (errorBoundaryComponent as any)?.displayName ||
+      errorBoundaryComponent?.name ||
+      'Unknown'
+
+    const componentThatErroredFrame = errorInfo?.componentStack?.split('\n')[1]
+
+    // Match chrome or safari stack trace
+    const matches =
+      componentThatErroredFrame?.match(/\s+at (\w+)\s+|(\w+)@/) ?? []
```

Comment:
Some examples here please to make it easier to scan what this is supposed to do.